network/newstream/stream.go

```diff
@@ -99,27 +131,825 @@ func (s *SlipStream) removePeer(p *Peer) {
+					p.logger.Trace("got chunk from peer", "addr", cc.Address(), "left", v)
+					if v == 0 {
+						p.logger.Debug("done receiving chunks for open want", "ruid", w.ruid)
+						close(errc)
+						return
+					}
+				}()
+			case <-p.quit:
+				return
+			case <-w.done:
+				return
+			case <-s.quit:
+				return
+			}
+		}
+	}()
+	return errc
+}
+func (s *SlipStream) serverCollectBatch(ctx context.Context, p *Peer, provider StreamProvider, key interface{}, from, to uint64) (hashes []byte, f, t uint64, empty bool, err error) {
+	p.logger.Debug("stream.CollectBatch", "from", from, "to", to)
+	batchStart := time.Now()
+
+	descriptors, stop := provider.Subscribe(ctx, key, from, to)
+	defer stop()
+
+	const batchTimeout = 500 * time.Millisecond
+
+	var (
+		batch        []byte
+		batchSize    int
+		batchStartID *uint64
+		batchEndID   uint64
+		timer        *time.Timer
+		timerC       <-chan time.Time
+	)
+
+	defer func(start time.Time) {
+		metrics.GetOrRegisterResettingTimer("stream.serverCollectBatch.total-time", nil).UpdateSince(start)
+		metrics.GetOrRegisterCounter("stream.serverCollectBatch.batch-size", nil).Inc(int64(batchSize))
+		if timer != nil {
+			timer.Stop()
+		}
+	}(batchStart)
+
+	for iterate := true; iterate; {
+		select {
+		case d, ok := <-descriptors:
+			if !ok {
+				iterate = false
+				break
+			}
+			s.logger.Trace("got address on subscribe", "address", d.Address.String(), "ident", batchStart, "key", key)
+			batch = append(batch, d.Address[:]...)
+			batchSize++
+			if batchStartID == nil {
+				// set batch start id only if
+				// this is the first iteration
+				batchStartID = &d.BinID
+			}
+			batchEndID = d.BinID
+			if batchSize >= BatchSize {
+				iterate = false
+				metrics.GetOrRegisterCounter("stream.serverCollectBatch.full-batch", nil).Inc(1)
+				p.logger.Trace("pull subscription - batch size reached", "batchSize", batchSize, "batchStartID", *batchStartID, "batchEndID", batchEndID)
+			}
+			if timer == nil {
+				timer = time.NewTimer(batchTimeout)
+			} else {
+				if !timer.Stop() {
+					<-timer.C
+				}
+				timer.Reset(batchTimeout)
+			}
+			timerC = timer.C
+		case <-timerC:
+			// return batch if new chunks are not received after some time
+			iterate = false
+			metrics.GetOrRegisterCounter("stream.serverCollectBatch.timer-expire", nil).Inc(1)
+			p.logger.Trace("pull subscription timer expired", "batchSize", batchSize, "batchStartID", batchStartID, "batchEndID", batchEndID)
+		case <-p.quit:
+			iterate = false
+			p.logger.Trace("pull subscription - quit received", "batchSize", batchSize, "batchStartID", batchStartID, "batchEndID", batchEndID)
+		case <-s.quit:
+			iterate = false
+			p.logger.Trace("pull subscription - shutting down")
+		}
+	}
+	if batchStartID == nil {
+		// if batch start id is not set, it means we timed out
+		return nil, 0, 0, true, nil
+	}
+	return batch, *batchStartID, batchEndID, false, nil
+}
+
+func (s *SlipStream) PeerCursors() string {
```

Comment:
It appears that you are adding this API endpoint for humans to query over Websocket for debug purposes.

It might be helpful if this returns a structured JSON output, so that this API is not only human readable, but also we can use it in integration tests, if this even makes sense.

Just an idea.